### ISSUE TITLE
[c++] unable to expand recursive macro

### ISSUE THREAD

**Issue by ZSaberLv0** [doxygen/doxygen#8448]
**To Reproduce**
[DoxygenTest.zip](https://github.com/doxygen/doxygen/files/6176850/DoxygenTest.zip)

the main code pasted here:

``` cpp
#define EMPTY(...)
#define NO_DEFAULT_PARAM() EMPTY
#define DEFAULT_PARAM(v) = v EMPTY

#define METHOD(name, def) \
    /** @brief doc for name */ \
    extern void name(int param def());

METHOD(f0, NO_DEFAULT_PARAM())
METHOD(f1, DEFAULT_PARAM(1))
```

use the above files, simply run `doxygen`, and the result:
![image](https://user-images.githubusercontent.com/6846867/111894045-6f9d8180-8a42-11eb-921c-01099450f0fc.png)

**Version**
`1.8.17` was the first release version that reproduce the issue

**Comment by albert-github** [comment]
It looks like this comes from the commit:
```
Commit: 5fe991236c9042a1ea75d6693aca41b3256a0d89 [5fe9912]
Date: Monday, November 25, 2019 9:02:53 PM
Fixed preprocessor issue that resulted in "More #endif's than #if's found." warning.
```

Though the issue comment does not really give a clue that this has an influence on the macro substitution.
I did a quick test with the current master and the current master (1.9.2 (39d20a3660942b12f3b003d0986794568e86d984)) gives the problem, when undoing the specific fix of the commit the shown code is OK.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.2.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
